db/crud.go

```diff
@@ -882,278 +1072,109 @@ func (db *Database) IsIllegalConflict(doc *Document, parentRevID string, deleted
-//      On cas failure, the document will still be reloaded from the bucket as usual.
-func (db *Database) updateAndReturnDoc(
-	docid string,
-	allowImport bool,
-	expiry uint32,
-	existingDoc *sgbucket.BucketDocument, // If existing is present, passes these to WriteUpdateWithXattr to allow bypass of initial GET
-	callback updateAndReturnDocCallback) (docOut *Document, newRevID string, err error) {
-	key := realDocID(docid)
-	if key == "" {
-		return nil, "", base.HTTPErrorf(400, "Invalid doc ID")
-	}
+	var storedBody Body
+	var syncFunctionExpiry *uint32
 
-	// Added annotation to the following variable declarations for reference during future refactoring of documentUpdateFunc into a standalone function
-	var doc *Document                                // Passed to documentUpdateFunc as pointer, may be possible to define in documentUpdateFunc
-	var body Body                                    // Could be returned by documentUpdateFunc
-	var storedBody Body                              // Persisted revision body, used to update rev cache
-	var changedPrincipals, changedRoleUsers []string // Could be returned by documentUpdateFunc
-	var docSequence uint64                           // Must be scoped outside callback, used over multiple iterations
-	var unusedSequences []uint64                     // Must be scoped outside callback, used over multiple iterations
-	var oldBodyJSON string                           // Could be returned by documentUpdateFunc.  Stores previous revision body for use by DocumentChangeEvent
-
-	// documentUpdateFunc applies the changes to the document.  Called by either WriteUpdate or WriteUpdateWithXATTR below.
-	documentUpdateFunc := func(doc *Document, docExists bool, importAllowed bool) (updatedDoc *Document, writeOpts sgbucket.WriteOptions, updatedExpiry *uint32, err error) {
+	err = db.preCallbackValidation(doc, allowImport, docExists)
+	if err != nil {
+		return doc, syncFunctionExpiry, nil, ``, storedBody, ``, previousDocSequence, unusedSequences, changedPrincipals, changedRoleUsers, err
+	}
 
-		var newAttachments AttachmentData
-		// Be careful: this block can be invoked multiple times if there are races!
-		if !importAllowed && docExists && !doc.HasValidSyncData(db.writeSequences()) {
-			err = base.HTTPErrorf(409, "Not imported")
-			return
-		}
+	// Invoke the callback to update the document and return a new revision body:
+	body, newAttachments, updatedExpiry, err := callback(doc)
+	if err != nil {
+		return doc, syncFunctionExpiry, body, ``, storedBody, ``, previousDocSequence, unusedSequences, changedPrincipals, changedRoleUsers, err
+	}
 
-		// Invoke the callback to update the document and return a new revision body:
-		body, newAttachments, updatedExpiry, err = callback(doc)
-		if err != nil {
-			return
-		}
+	err = db.postCallbackValidation(body)
```

Comment:
Similarly, postCallbackValidation is really 'validateNewBody' - checking that the new body provided by the callback is valid.